### ISSUE TITLE
Operator ignores PrometheusCompatibilityMatrix and installs `latest` prometheus

### ISSUE THREAD

**Issue by amavrin** [prometheus-operator/prometheus-operator#4237]
**What happened?**
Operator `v0.49.0` is installed. `Prometheus` objects is installed with no explicit version set.
Operator installs `latest` version of Prometheus image.

**Did you expect to see something different?**
I'd expect that the latest version from `PrometheusCompatibilityMatrix` is installed
https://github.com/prometheus-operator/prometheus-operator/blob/release-0.49/pkg/operator/defaults.go#L96

**How to reproduce it (as minimally and precisely as possible)**:
Create an empty cluster (e.g. in Katacoda) and execute
```
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.49/bundle.yaml

cat > prometheus.yaml << EOF
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: prometheus
spec:
  serviceMonitorSelector:
    matchLabels:
      team: frontend
  resources:
    requests:
      memory: 400Mi
EOF

kubectl apply -f prometheus.yaml
```
Wait for Prometheus pod to be created and check
```
kubectl get pod prometheus-prometheus-0 -o json | jq '.status.containerStatuses[] | select(.name=="prometheus") | .image'
```
The result is `quay.io/prometheus/prometheus:latest`

**Environment**
https://www.katacoda.com/courses/kubernetes/playground

* Prometheus Operator version:
`version=0.49.0`

* Kubernetes version information:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:50:46Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```
* Kubernetes cluster kind:
Cluster created at https://www.katacoda.com

* Prometheus Operator Logs:

```
level=info ts=2021-08-30T15:03:25.739668917Z caller=main.go:295 msg="Starting Prometheus Operator" version="(version=0.49.0, branch=master, revision=2388bfa55)"
level=info ts=2021-08-30T15:03:25.739776792Z caller=main.go:296 build_context="(go=go1.16.5, user=paulfantom, date=20210706-17:43:37)"
level=info ts=2021-08-30T15:03:25.815775169Z caller=main.go:115 msg="Starting insecure server on [::]:8080"
level=info ts=2021-08-30T15:03:25.832455877Z caller=operator.go:449 component=alertmanageroperator msg="connection established" cluster-version=v1.18.0
level=info ts=2021-08-30T15:03:25.832510918Z caller=operator.go:458 component=alertmanageroperator msg="CRD API endpoints ready"
level=info ts=2021-08-30T15:03:25.832715727Z caller=operator.go:309 component=thanosoperator msg="connection established" cluster-version=v1.18.0
level=info ts=2021-08-30T15:03:25.832739477Z caller=operator.go:318 component=thanosoperator msg="CRD API endpoints ready"
level=info ts=2021-08-30T15:03:25.832829203Z caller=operator.go:421 component=prometheusoperator msg="connection established" cluster-version=v1.18.0
level=info ts=2021-08-30T15:03:25.832849497Z caller=operator.go:430 component=prometheusoperator msg="CRD API endpoints ready"
level=info ts=2021-08-30T15:03:26.620274138Z caller=operator.go:261 component=thanosoperator msg="successfully synced all caches"
level=info ts=2021-08-30T15:03:26.708033233Z caller=operator.go:281 component=alertmanageroperator msg="successfully synced all caches"
level=info ts=2021-08-30T15:03:26.708420165Z caller=operator.go:351 component=prometheusoperator msg="successfully synced all caches"
level=info ts=2021-08-30T15:03:26.708736417Z caller=operator.go:1221 component=prometheusoperator key=default/prometheus msg="sync prometheus"
level=info ts=2021-08-30T15:03:27.114665806Z caller=operator.go:1221 component=prometheusoperator key=default/prometheus msg="sync prometheus"
level=info ts=2021-08-30T15:03:27.412587339Z caller=operator.go:1221 component=prometheusoperator key=default/prometheus msg="sync prometheus"
level=info ts=2021-08-30T15:03:49.999167607Z caller=operator.go:1221 component=prometheusoperator key=default/prometheus msg="sync prometheus"
```

**Anything else we need to know?**:
Probably it functions as designed, but it is not clear from the documentation.
Also, this results in that Operator installs Prometheus version that is not in its compatibility matrix.

**Comment by paulfantom** [comment]
IMHO since `PrometheusCompatibilityMatrix` is not really used for anything apart from docs generation and test, we should remove it, move to the appropriate place, or start using it somewhere :)

As for the relation between `version` and `image`, it can be confusing. The current flow is that `version` is just metadata and does not influence `image`. This is because `image` can be set in multiple ways and it would be error-prone to parse them all. However, we need `version` to figure out the feature set available for the specified Prometheus server.

What if we would parse `image` on a best-effort basis to figure out `version` and compare this with `PrometheusCompatibilityMatrix`? After this, we could put the additional field in `.Status` reflecting the observed Prometheus version (and thus feature set)?

---
On top of `version` and `image`, we also have deprecated fields like `baseImage`, `tag`, and `sha`. Plus we have problems with too large CRD definition. Maybe it is time to design CRD v2?

---
cc @fpetkovski @simonpasquier

**Comment by simonpasquier** [comment]
> What if we would parse image on a best-effort basis to figure out version and compare this with PrometheusCompatibilityMatrix? After this, we could put the additional field in .Status reflecting the observed Prometheus version (and thus feature set)?

My take is that the operator should only rely on the `spec.Version` field to infer the Prometheus version. Users can rebuild their own images and use a different version scheme, or they can reference images by SHA digest hence trying to parse the version from the image location seems fragile.

Overall I would expect that when a user creates a minimal Prometheus resource (as @amavrin did), the operator picks up the upstream image that maps to the latest item in `PrometheusCompatibilityMatrix`.

**Comment by antoinedeschenes** [comment]
What if `DefaultPrometheusVersion` was used by default? Specifying `latest` in the `--prometheus-default-base-image` flag would bring back the current behavior. (Although it seems that a default image tag would override `p.Spec.Version`) https://github.com/prometheus-operator/prometheus-operator/blob/036e7c4437e848822fd66e3ed0a3652e360db2f7/pkg/operator/image.go#L38-L42

**Comment by simonpasquier** [comment]
closed by #5171
